### ISSUE TITLE
[Find2.0][MIGraphX] Quant Convolution fails with Find 2.0

### ISSUE THREAD

**Issue by umangyadav** [ROCm/MIOpen#1800]
For the quant convolution output tensor descriptor is of type int32 and innputs are either int8 or int8x4. 

Find2.0 uses "MIOpenSubTensorOpWithScalarKernel.cl" program which doesn't support int32 type. 

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/5e78e715b174edc2119560f41dcababeb1957fac/src/ocl/tensorocl.cpp#L1438

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/5e78e715b174edc2119560f41dcababeb1957fac/src/kernels/MIOpenSubTensorOpWithScalarKernel.cl#L55

I was able to make MIGraphX quant convolution work by adding INT32 type in that file.

```
#if MIOPEN_USE_INT32 == 1
#define _FLOAT int
#ifndef FLT_MAX 2,147,483,647
#define MAX_VAL 
#else 
#define MAX_VAL FLT_MAX
#endif
#endif
```

Even though it  works for me, i am not sure if this is the right fix. Can we get proper fix in MIOpen to handle this case ?

**Comment by junliume** [comment]
@DrizztDoUrden @atamazov can we take a quick look at this issue?

**Comment by atamazov** [comment]
@umangyadav Does it work with Find 1.0 without your fix? If yes, then it is @DrizztDoUrden who can provide the proper fix soon.

Can we reproduce the issue using our tests or driver? Easy reproduction is important for speed.

**Comment by umangyadav** [comment]
@atamazov Yes it works with find1.0 without my fix.

**Comment by umangyadav** [comment]
@atamazov For the production, i am not much aware with unit-tests or driver  on MIOpen side. Does miopen-driver suppport find.2.0 ? 

Otherwise, i can locally test it with MIGraphX once the PR is up.

**Comment by atamazov** [comment]
@umangyadav 

> @atamazov For the production, i am not much aware with unit-tests or driver on MIOpen side. Does miopen-driver suppport find.2.0 ?

In #1503 (that you have partially reviewed) you can see new Find2.0 specific tests (so now we have `test_con2d_find2` and `test_conv3d_find2`). If you have difficulties with those (e.g. I am not sure that INT32 is supported), just let us know failing convolution configs. You can use MIOpenDriver's syntax as if you use Find 1.0. Or whatever format that you like.

> Otherwise, i can locally test it with MIGraphX once the PR is up.

Thanks. But the _process_ we are using implies adding regression tests together with a fix. We need this not to only reproduce issue but also to add a regression test.

**Comment by umangyadav** [comment]
I tried reproducing now. But it didn't.  It is passing with the latest develop branch build. 

I was comparing the MIOpen log from both cases and I see that in the failure case: it is looking for kernel in `dirty` kernel cache. 

Failure log : 
[quant_conv_failure.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9735396/quant_conv_failure.txt.txt)

Passing Log : 
[quant_conv_passing.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9735397/quant_conv_passing.txt)

Could that be why it was failing for me before?

**Comment by umangyadav** [comment]
@atamazov @DrizztDoUrden 

Here is the minimal example, i could come up with. 

[CMakeLists.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9759523/CMakeLists.txt)
[load_and_save.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9759528/load_and_save.txt)

Steps to reproduce : 

1. rename `load_and_save.txt` to `load_and_save.cpp` ( as i couldn't attach .cpp file). 
2. Use cmakelists.txt to build. It assumes rocm and miopen are available at /opt/rocm. 
3. `mkdir build && cd build &&  cmake .. && make -j$(nproc)`
4. `export MIOPEN_DISABLE_CACHE=1`
5. `./load_and_save`

**Comment by umangyadav** [comment]
> @atamazov @DrizztDoUrden
> 
> Here is the minimal example, i could come up with.
> 
> [CMakeLists.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9759523/CMakeLists.txt) [load_and_save.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9759528/load_and_save.txt)
> 
> Steps to reproduce :
> 
> 1. rename `load_and_save.txt` to `load_and_save.cpp` ( as i couldn't attach .cpp file).
> 2. Use cmakelists.txt to build. It assumes rocm and miopen are available at /opt/rocm.
> 3. `mkdir build && cd build &&  cmake .. && make -j$(nproc)`
> 4. `export MIOPEN_DISABLE_CACHE=1`
> 5. `./load_and_save`

@DrizztDoUrden @atamazov Are you able to reproduce the error with this test case ?

**Comment by atamazov** [comment]
[side note] @umangyadav please avoid excessive quotation.

**Comment by atamazov** [comment]
@umangyadav Which GPU shall we use?

**Comment by umangyadav** [comment]
I was using MI100, but any would do i think.

**Comment by atamazov** [comment]
@umangyadav Reproduced on Navi21 with `test_con2d_find2` (but it requires some patch to support int8). FYI command to reproduce:

```bash
MIOPEN_LOG_LEVEL=6 \
MIOPEN_FIND_MODE=1 \
./bin/test_conv2d_find2 --int8 --cmode conv --pmode default --group-count 1 \
--batch_size 2 --input_channels 3 --output_channels 2 --spatial_dim_elements 4 4 \
--filter_dims 3 3 --pads_strides_dilations 2 0 2 1 2 1 \
--disable-backward-data --disable-backward-weights \
--verbose
```

> Even though it works for me, i am not sure if this is the right fix. Can we get proper fix in MIOpen to handle this case ?

The fix should not affect kernels, so most likely your fix is not the proper one.

I am investigating; stay tuned.

@DrizztDoUrden This is the patch that makes `--int8` working with test_con2d_find2: [test_conv2d_fin2__int8_fix.diff.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/9790590/test_conv2d_fin2__int8_fix.diff.txt). Is it correct, what do you think?

Please look into the issue if you have spare time (not very urgent but a couple of extra eyes would be nice)

**Comment by atamazov** [comment]
~@DrizztDoUrden The problem happens somewhere in `GemmFwdRest` where int8 type is not handled properly~

**Comment by atamazov** [comment]
Error happens during SetTensor call, https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1503/files#r996216187

@DrizztDoUrden Find 1.0 does not use it and I think it should not be there. Let's make a quick call when you have time.

**Comment by atamazov** [comment]
@umangyadav The root reason is questionable API "feature", see here: (https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1503/files#r996219686). Let's discuss it in some dedicated ticket.

**Comment by atamazov** [comment]
@umangyadav 
>> Even though it works for me, i am not sure if this is the right fix. Can we get proper fix in MIOpen to handle this case ?
> 
> The fix should not affect kernels, so most likely your fix is not the proper one.

UPDATE: Your fix adds support for INT32, which is necessary if we decide to keep current implementation of the host code. Maybe we shall use `unsigned` instead of `int` but otherwise it looks good.

**Comment by atamazov** [comment]
I will provide ~W/A~ some solution for this tomorrow.

**Comment by umangyadav** [comment]
@atamazov Thanks for investigating. 
Can we get a temporary workaround in 5.4?

**Comment by DrizztDoUrden** [comment]
I would say that there are two main ways of solving this with different purposes in mind.
The first is to add support for the type to the kernel as you have suggested. I don't really see the downsides. A benefit of doing it first is that you won't have to change the way you are using the library.
The second is to allow the consumer to provide his own buffers as in find 1.0, but via find options. I don't see cons of implementing either. This also would have additional benefit. But specifically for MIGraphX that alone would require some changes to how the library is used. If MIGraphX actually allocates tensors before the find, this is probably the preferred way to remove duplicate memory allocation. Although I am not sure it should allocate them before find if the library is able to properly handle that. But that depends on the use cases.
Both could and maybe should be implemented simultaneously. Do you have any preference from the MIGraphX perspective on what would you like to have first?

**Comment by atamazov** [comment]
> @atamazov Thanks for investigating. Can we get a temporary workaround in 5.4?

I think yes. Let's ask @junliume to cherry-pick ~W/A~ necessary change into 5.4 release branch.

**Comment by atamazov** [comment]
@umangyadav #1828 opened.

**Comment by atamazov** [comment]
@umangyadav Please comment on https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1800#issuecomment-1280864283. Current Find 2.0 implementation simplifies the user's apps but the cost is loss of flexibility. Also the GPU memory is now owned by two: the library (which needs to allocate buffers for Find call) and the user's app, -- this is not good (especially if tensors are very big).

**Comment by atamazov** [comment]
@umangyadav oops, now I see your comment at https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1503#discussion_r997059225

**Comment by umangyadav** [comment]
>On the memory allocation 

Currently, MIGraphX is not allocating any buffers itself for calling `Find-2.0`. It is just passing the tensor descriptors. But it is allocating buffers for Find-Immediate API and  passing them through `miopenFindConvolutionForwardAlgorithm`. So, I think it is not duplicating buffer allocations for `Find` process. 

FYI @pfultz2

**Comment by atamazov** [comment]
@umangyadav oops, now I see your comment at https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1503#discussion_r997059225

**Comment by ppanchad-amd** [comment]
@umangyadav Is this fixed with latest ROCm 6.1.0? If resolved, please close ticket. Thanks!

**Comment by umangyadav** [comment]
yes this is an old issue. Fixed now.
